File: AppRemover/src/main/java/com/vs_unusedappremover/MyApplication.java
package com.vs_unusedappremover;

import android.app.Application;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageManager;
import android.content.pm.PackageManager.NameNotFoundException;
import android.os.Handler;

import com.squareup.picasso.Picasso;
import com.vs_unusedappremover.common.GA;
import com.vs_unusedappremover.data.ApplicationCollection;
import com.vs_unusedappremover.data.DatabaseHelper;

import java.io.File;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.concurrent.ThreadPoolExecutor;
import java.util.concurrent.TimeUnit;

public class MyApplication extends Application {
	
	public static final String TAG = MyApplication.class.getSimpleName();
	private static final String INSTALL_TIME = "install time";
	
	private static MyApplication instance;
	
	private DatabaseHelper database;
	private final ThreadPoolExecutor backgroundExecutor;
	private Handler uiHandler;
	private ApplicationCollection applications;
	private Picasso picasso;

	public static MyApplication getInstance() {
		return instance;
	}
	
	public static void startMonitoringService(Context context) {
		context.startService(new Intent(context, MonitorApplicationsService.class));
	}
	
	public MyApplication() {
		backgroundExecutor = new ThreadPoolExecutor(2, 16, 1, TimeUnit.SECONDS, new LinkedBlockingQueue<Runnable>());
		backgroundExecutor.setRejectedExecutionHandler(new ThreadPoolExecutor.CallerRunsPolicy());
	}
		
	@Override
	public void onCreate() {
		super.onCreate();
		instance = this;
		uiHandler = new Handler();
        GA.init(null, this);
		applications = new ApplicationCollection(this);
		picasso = new Picasso.Builder(this)
				.addRequestHandler(AppIcon.createRequestHandler(this))
				.build();
		startMonitoringService(this);
		checkInstallTimePresent();
	}
	
	public synchronized DatabaseHelper getDbHelper() {
		if (database == null) {
			database = new DatabaseHelper(this);
		}
		return database;
	}
	
	public ExecutorService getBackgroundExecutor() {
		return backgroundExecutor;
	}
	
	public ApplicationCollection getApplications() {
		return applications;
	}
	
	public Handler getUiHandler() {
		return uiHandler;
	}
	
	public long getInstallTime() {
		SharedPreferences prefs = getSharedPreferences(TAG, MODE_PRIVATE);
		return prefs.getLong(INSTALL_TIME, 0);
	}

	public Picasso picasso() {
		return picasso;
	}
	
	private void checkInstallTimePresent() {
		SharedPreferences prefs = getSharedPreferences(TAG, MODE_PRIVATE);
		if (prefs.getLong(INSTALL_TIME, 0) == 0) {
			prefs.edit()
				.putLong(INSTALL_TIME, calculateInstallTime())
				.apply();
		}
	}	
	
	private long calculateInstallTime() {
		long now = System.currentTimeMillis();
		PackageManager pm = getPackageManager();
		try {
			ApplicationInfo info = pm.getApplicationInfo(getPackageName(), 0);
			File apk = new File(info.sourceDir);
			long installTime = apk.lastModified();
			return installTime != 0 ? installTime : now;
		} catch (NameNotFoundException e) {
			return now;
		}
	}
}

File: AppRemover/src/main/java/com/vs_unusedappremover/MyApplication.kt
package com.vs_unusedappremover

import android.app.Application
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager.NameNotFoundException
import android.os.Handler
import com.squareup.picasso.Picasso
import com.vs_unusedappremover.common.GA
import com.vs_unusedappremover.data.ApplicationCollection
import com.vs_unusedappremover.data.DatabaseHelper
import java.io.File
import java.util.concurrent.ExecutorService
import java.util.concurrent.LinkedBlockingQueue
import java.util.concurrent.ThreadPoolExecutor
import java.util.concurrent.TimeUnit

class MyApplication : Application() {

    val dbHelper: DatabaseHelper by lazy {
        DatabaseHelper(this)
    }

    val backgroundExecutor: ExecutorService
    lateinit var uiHandler: Handler
        private set
    lateinit var applications: ApplicationCollection
        private set
    lateinit private var picasso: Picasso

    init {
        backgroundExecutor = ThreadPoolExecutor(2, 16, 1, TimeUnit.SECONDS, LinkedBlockingQueue())
        backgroundExecutor.rejectedExecutionHandler = ThreadPoolExecutor.CallerRunsPolicy()
    }

    override fun onCreate() {
        super.onCreate()
        instance = this
        uiHandler = Handler()
        GA.init(null, this)
        applications = ApplicationCollection(this)
        picasso = Picasso.Builder(this)
                .addRequestHandler(AppIcon.createRequestHandler(this))
                .build()
        startMonitoringService(this)
        checkInstallTimePresent()
    }

    val installTime: Long
        get() {
            val prefs = getSharedPreferences(TAG, Context.MODE_PRIVATE)
            return prefs.getLong(INSTALL_TIME, 0)
        }

    fun picasso(): Picasso? {
        return picasso
    }

    private fun checkInstallTimePresent() {
        val prefs = getSharedPreferences(TAG, Context.MODE_PRIVATE)
        if (prefs.getLong(INSTALL_TIME, 0) == 0L) {
            prefs.edit()
                    .putLong(INSTALL_TIME, calculateInstallTime())
                    .apply()
        }
    }

    private fun calculateInstallTime(): Long {
        val now = System.currentTimeMillis()
        val pm = packageManager
        try {
            val info = pm.getApplicationInfo(packageName, 0)
            val apk = File(info.sourceDir)
            val installTime = apk.lastModified()
            return if (installTime != 0L) installTime else now
        } catch (e: NameNotFoundException) {
            return now
        }

    }

    companion object {

        val TAG: String = MyApplication::class.java.simpleName
        private val INSTALL_TIME = "install time"

        lateinit var instance: MyApplication
            private set

        fun startMonitoringService(context: Context) {
            context.startService(Intent(context, MonitorApplicationsService::class.java))
        }
    }
}
